src/ReservoirPriceOracle.sol

```diff
@@ -1,32 +1,405 @@
+    //                                       MODIFIERS                                           //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    modifier validateTokens(address aTokenA, address aTokenB) {
+        if (aTokenA == aTokenB) revert RPC_SAME_TOKEN();
+        _;
+    }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                   PUBLIC FUNCTIONS                                        //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    // IPriceOracle
+
+    /// @inheritdoc IPriceOracle
+    function getQuote(uint256 aAmount, address aBase, address aQuote)
+        external
+        view
+        validateTokens(aBase, aQuote)
+        returns (uint256 rOut)
+    {
+        rOut = _getQuote(aAmount, aBase, aQuote);
+    }
+
+    /// @inheritdoc IPriceOracle
+    function getQuotes(uint256 aAmount, address aBase, address aQuote)
+        external
+        view
+        validateTokens(aBase, aQuote)
+        returns (uint256 rBidOut, uint256 rAskOut)
+    {
+        uint256 lResult = _getQuote(aAmount, aBase, aQuote);
+        (rBidOut, rAskOut) = (lResult, lResult);
+    }
+
+    // price update related functions
+
+    function gasBountyAvailable() external view returns (uint256) {
+        return address(this).balance;
+    }
+
+    function route(address aToken0, address aToken1) external view returns (address[] memory) {
+        return _route[aToken0][aToken1];
+    }
+
+    /// @notice Updates the TWAP price for all simple routes between `aTokenA` and `aTokenB`. Will also update intermediate routes if the route defined between
+    /// aTokenA and aTokenB is longer than 1 hop
+    /// However, if the route between aTokenA and aTokenB is composite route (more than 1 hop), no cache entry is written
+    /// for priceCache[aTokenA][aTokenB] but instead the prices of its constituent simple routes will be written.
+    /// @param aTokenA Address of one of the tokens for the price update. Does not have to be less than address of aTokenB
+    /// @param aTokenB Address of one of the tokens for the price update. Does not have to be greater than address of aTokenA
+    /// @param aRewardRecipient The beneficiary of the reward. Must implement the receive function if is a smart contract address
+    function updatePrice(address aTokenA, address aTokenB, address aRewardRecipient)
         external
+        validateTokens(aTokenA, aTokenB)
+        nonReentrant
+    {
+        (address lToken0, address lToken1) = aTokenA.sortTokens(aTokenB);
+
+        address[] memory lRoute = _route[lToken0][lToken1];
+        if (lRoute.length == 0) revert PO_NoPath();
+
+        OracleAverageQuery[] memory lQueries = new OracleAverageQuery[](lRoute.length - 1);
```

Comment:
Is there a benefit to allocating the queries in a memory array or can we just dispatch the calls immediately and avoid this allocation?

Reply:
1. I've attempted many times in the past and recently to initialize this in-memory array within the function call itself but it seems that solidity doesn't allow that
1. Right now having this temporary variable serves another purpose, which is that it stores the base and quote in the sorted sequence, so that after we get the results back from `getWeightedAverage` we don't need an operation to figure out which is base and quote again